### PR TITLE
config: use fields to detect enterprise-only settings

### PR DESCRIPTION
**Pull request by dnephin**
Now that config can come from non-file sources it is not safe to assumes that the mapstructure `Metdata.Keys` will contain all the keys because some config can be specified with command line flags or literals.

This change allows us to remove the json marshal/unmarshal cycle for command line flags, which will allow us to remove all of the hcl/json struct tags on config fields. That is done in #9250 which is a follow up to this PR.

This PR also does a bit of cleanup in some of the config tests to remove unused things and make the failure messages easier to understand.

### CODE REVIEW

File: agent/config/builder_oss.go [hunk 1]
```diff
@@ -4,47 +4,55 @@ package config
...
+// for enterprise configurations used in OSS.
+func validateEnterpriseConfigKeys(config *Config) []error {
+	var result []error
+	add := func(k string) {
+		result = append(result, enterpriseConfigKeyError{key: k})
 	}
-)
+
+	if config.ReadReplica != nil {
+		add("read_replica")
```
Comment (dnephin) [inline agent/config/builder_oss.go:20]: This should say `"read_replica" or "non_voting_server"` to handle the deprecated config key.

File: agent/config/builder.go [hunk 2]
```diff
@@ -1852,10 +1845,18 @@ func (b *Builder) stringValWithDefault(v *string, defaultVal string) string {
 	return *v
 }
 
+// Deprecated: use stringVal
```
Comment (s-christoff) [inline agent/config/builder.go:1848]: This is a bit confusing unless you look down a bit, which wasn't my first intention. It's a silly nit, but this might benefit from being reworded to have something like "Please look a couple lines down" perhaps?

File: agent/config/runtime_test.go [hunk 3]
```diff
@@ -4957,12 +4940,19 @@ func testConfig(t *testing.T, tests []configTest, dataDir string) {
...
 					tt.patchActual(&actual)
 				}
-				require.Equal(t, expected, actual)
+				assertDeepEqual(t, expected, actual, cmpopts.EquateEmpty())
 			})
 		}
 	}
 }
 
+func assertDeepEqual(t *testing.T, x, y interface{}, opts ...cmp.Option) {
```
Comment (s-christoff) [inline agent/config/runtime_test.go:4949]: We discussed putting this is a more centralized place - maybe best to do it in this PR first?

**Review by s-christoff — APPROVED** [review]
Few minor comments but LGTM

File: agent/config/builder.go (hunk 2 above)
Comment (dnephin) [reply agent/config/builder.go:1848]: Ah yes, because both the Builder method and the function are named the same thing, that is confusing! I'll reword this to clarify.
